beluga_tutorial/src/main.cpp

```diff
@@ -0,0 +1,258 @@
+      node["simulation_dataset"][cycle]["particles"]["reweight"]["weights"].push_back(
+          static_cast<double>(std::get<1>(reweight)));
+    }
+
+    for (const auto& resample : data.resample) {
+      node["simulation_dataset"][cycle]["particles"]["resample"]["states"].push_back(std::get<0>(resample));
+      node["simulation_dataset"][cycle]["particles"]["resample"]["weights"].push_back(
+          static_cast<double>(std::get<1>(resample)));
+    }
+
+    node["simulation_dataset"][cycle]["estimation"]["mean"] = std::get<0>(data.estimation);
+    node["simulation_dataset"][cycle]["estimation"]["sd"] = std::get<1>(data.estimation);
+  }
+
+  const std::string bags_path{"./src/beluga/beluga_tutorial/bags/"};
+  std::filesystem::create_directory(bags_path);
+  std::ofstream fout(bags_path + file_name, std::ios::trunc);
+  fout << node;
+  fout.close();
+}
+
+int main() {
+  // Load params from yaml file
+  TutorialParams tutorial_params;
+  load_params_from_yaml(tutorial_params);
+
+  // Create the TutorialDataset object for record the simulation data
+  TutorialDataset tutorial_dataset;
+
+  // Load map from file or generate a random map
+  LandmarkMapVector landmark_map;
+  if (tutorial_params.load_landmark_map_from_file) {
+    load_landmark_map_from_yaml(landmark_map);
+  } else {
+    std::uniform_int_distribution landmark_distribution{0, tutorial_params.map_size};
+    landmark_map = beluga::views::sample(landmark_distribution) |
+                   ranges::views::take_exactly(tutorial_params.number_of_doors) | ranges::to<LandmarkMapVector>;
+    landmark_map |= ranges::actions::sort | ranges::actions::unique;
+  }
+
+  tutorial_dataset.landmark_map = landmark_map;
+
+  // Generate particles
+  std::normal_distribution<double> initial_distribution(tutorial_params.initial_pose, tutorial_params.initial_pose_sd);
+  auto particles = beluga::views::sample(initial_distribution) |
+                   ranges::views::transform(beluga::make_from_state<Particle>) |
+                   ranges::views::take_exactly(tutorial_params.number_of_particles) | ranges::to<beluga::TupleVector>;
+
+  // Execute the particle filter using beluga
+  double current_pose{tutorial_params.initial_pose};
+  for (auto n = 0; n < tutorial_params.number_of_cycles; n++) {
+    // Prepare the data to be saved
+    TutorialData tutorial_data;
+
+    // Save ground truth
+    current_pose += (tutorial_params.velocity * tutorial_params.dt);
+    tutorial_data.ground_truth = current_pose;
+
+    // Check if the simulation is out of bounds
+    if (current_pose > tutorial_params.map_size) {
+      break;
+    }
+
+    // Motion model
+    auto motion_model = [&](double state) {
+      double distance = (tutorial_params.velocity * tutorial_params.dt);
+      double translation_param = generateRandom(0.0, tutorial_params.translation_sigma);
+      return state + distance - translation_param;
+    };
```

Comment:
@Alondruck consider inlining `generateRandom` and letting Beluga give you a PRNG:
```suggestion
    auto motion_model = [&](double state, auto& gen) {
      const double distance = (tutorial_params.velocity * tutorial_params.dt);
      const std::normal_distribution<double> distribution(distance, tutorial_params.translation_sigma);
      return state + distribution(gen);
    };
```

Reply:
This way it looks better